### DISCUSSION TITLE
Writing unit tests with a service account

### DISCUSSION THREAD

**Discussion by TastyToasty** [pnp/pnpcore#622] [Q&A] [answered]
Hello,

i am trying to write some unit tests using pnp core. I just want to use a service account to do some read/write actions to a share point site. Are there any examples especially on how to handle the authentication?

Thanks in advance!

**Comment by jansenbe** [comment, marked as answer]
@TastyToasty : maybe our test projects can serve as inspiration for your unit tests as well, see https://github.com/pnp/pnpcore/tree/dev/src/sdk/PnP.Core.Test and https://github.com/pnp/pnpcore/tree/dev/src/sdk/PnP.Core.Admin.Test as examples. Our test project setup is slightly more complicated as we have an offline test concept, meaning you can run a test and record the server responses. You can then reply these responses to run your test fully offline. We've built this concept to allow for fast test execution while developing. See https://pnp.github.io/pnpcore/contributing/writing%20tests.html to learn more about how to use our testing system.

In general you would want to create a PnPContext in your test project and doing that is described here: https://pnp.github.io/pnpcore/using-the-sdk/readme.html. In your settings file you can chose to use the `CredentialManagerAuthenticationProvider` or `UsernamePasswordAuthenticationProvider` when you want to use a user+pwd to authenticate to SPO. See https://pnp.github.io/pnpcore/using-the-sdk/configuring%20authentication.html#using-the-credential-manager

**Reply by TastyToasty** [reply]
Thanks! Took me a while, but I figured it out with your help!
